Threaded block signing 📦

jgiszczak opened the pull request:
<!-- PLEASE FILL OUT THE FOLLOWING MARKDOWN TEMPLATE -->
<!-- PR title alone should be sufficient to understand changes. -->
## Change Description
Thread block signing.

## Change Type
**Select *ONE*:**
- [ ] Documentation
<!-- checked [x] = Documentation; unchecked [ ] = no changes, ignore this section -->
- [ ] Stability bug fix
<!-- checked [x] = Stability bug fix; unchecked [ ] = no changes, ignore this section -->
- [x] Other
<!-- checked [x] = Other; unchecked [ ] = no changes, ignore this section -->
- [ ] Other - special case
<!-- checked [x] = Other - special case; unchecked [ ] = no changes, ignore this section -->
<!-- Other - special case is for when a change warrants additional explanation or description in the release notes. Please include a description of the change for inclusion in the release notes. -->


## Testing Changes
**Select *ANY* that apply:**
- [x] New Tests
<!-- checked [x] = new test cases were added; unchecked [ ] = no new test cases -->
- [x] Existing Tests
<!-- checked [x] = existing test cases were edited; unchecked [ ] = no existing tests were modified -->
- [ ] Test Framework
<!-- checked [x] = this modifies the test framework; unchecked [ ] = no test framework changes -->
- [ ] CI System
<!-- checked [x] = this changes the CI system; unchecked [ ] = no CI changes -->
- [ ] Other
<!-- checked [x] = this integrates an external test system; unchecked [ ] = no miscellaneous test-related changes -->
<!-- Please describe your test changes, or list each new test and its purpose, under each respective checkbox -->


## Consensus Changes
- [ ] Consensus Changes
<!-- checked [x] = Consensus changes; unchecked [ ] = no changes, ignore this section -->
<!-- If this PR introduces a change to the validation of blocks in the chain or consensus in general, please describe the impact. -->


## API Changes
- [ ] API Changes
<!-- checked [x] = API changes; unchecked [ ] = no changes, ignore this section -->
<!-- If this PR introduces API changes, please describe the changes here. What will developers need to know before upgrading to this version? -->


## Documentation Additions
- [ ] Documentation Additions
<!-- checked [x] = Documentation changes; unchecked [ ] = no changes, ignore this section -->
<!-- Describe what must be added to the documentation after merge. -->

jgiszczak committed:
Thread block signing.  WIP

jgiszczak committed:
Properly copy prev activated protocol features and signing lambda

Remove debug code.

jgiszczak committed:
Remove unused enable shared from this.

jgiszczak committed:
Remove commented code.

jgiszczak committed:
Remove commented code from tests.

jgiszczak committed:
Carry a copy of signature provider vector with producer signing lambda

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2592,20 +2592,51 @@ block_state_ptr controller::finalize_block( const signer_callback_type& signer_c
 
    my->pending->_block_stage = completed_block{ bsp };
 
+   if( my->block_sign_fut.valid() ) {
```
This is not happening until the next block. I don't think we want to delay emitting the signed block by 500ms. This all should happen as soon as the block is signed. Can all this be moved into the lambda passed to block_sign_pool?

jgiszczak replied:
The signal emission is thread-safe but `log_irreversible()` uses `boost::multi_index_container` in the fork database which is not remotely thread-safe.  Emitting the block to the network early results in the peculiar case of the network possibly having an irreversible block which the local node does not consider irreversible for hundreds of milliseconds.  If a health monitor in a single producer network happens to shut down and restart a producing node in that window without activating a standby producer, the network will irreconcilably fork as the active producer essentially double-produces a block, having lost its own prior last block.

heifner replied:
I'm envisioning it working as follows:
- kick off a lambda to sign and do whatever else is possible in a diff thread
- When signing done it will post back to the main thread at high priority to finish out the `log_irreversible()` and `emit`.
- Indicate block is complete and pause on the next block at that location if the previous is still signing (should not happen in practice as it should not take more than 500ms to sign, although we should setup a test that causes it to happen by having a very slow sign proces..)

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2592,20 +2592,51 @@ block_state_ptr controller::finalize_block( const signer_callback_type& signer_c
 
    auto& ab = std::get<assembled_block>(my->pending->_block_stage);
 
+   auto pfa = *ab._pending_block_header_state.prev_activated_protocol_features.get();
```
`prev_activated_protocol_features` is a shared_ptr, why not just pass the shared_ptr around?

jgiszczak replied:
Done.

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -2346,7 +2346,7 @@ void producer_plugin_impl::produce_block() {
 
    //idump( (fc::time_point::now() - chain.pending_block_time()) );
-   block_state_ptr pending_blk_state = chain.finalize_block( [&]( const digest_type& d ) {
+   block_state_ptr pending_blk_state = chain.finalize_block( [relevant_providers]( const digest_type& d ) {
```
`relevant_providers` can be moved into the lambda.

jgiszczak replied:
Done.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2592,20 +2592,51 @@ block_state_ptr controller::finalize_block( const signer_callback_type& signer_c
 
```
`finalize_block` should take `signer_callback_type` by value.

jgiszczak replied:
Done.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2592,20 +2592,51 @@ block_state_ptr controller::finalize_block( const signer_callback_type& signer_c
+
+   my->block_sign_fut = async_thread_pool( my->block_sign_pool.get_executor(),
+                                           [pfa, pfs, bsp, signer_callback]() mutable {
+                                              (*bsp).sign( bsp->block, signer_callback );
```
signer_callback should be moved into the lambda

jgiszczak replied:
Done.

heifner commented on libraries/chain/block_header_state.cpp:
```diff
@@ -406,6 +401,8 @@ namespace eosio { namespace chain {
 
       verify_signee();
+
+      h->producer_signature = header.producer_signature;
```
Why pass in signed_block_header for the same object? This should have already been set on line 398.

huangminghuang replied:
The `h` here is coming from producer_plugin.cpp:2637 which is actually a member of `block_state`, so they are different. However, it seems unnecessary to convert from `signed_block_ptr` to `std::share_ptr<signed_block_header>` for the declaration of `h`.

jgiszczak replied:
`block_header_state.hpp` is unaware of the `signed_block` type.  It only knows about `signed_block_header`, for which a typedef/using shared pointer type doesn't exist.

heifner replied:
Then have it take a `signed_block_header& h` instead.

jgiszczak committed:
Various optimizations.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2585,27 +2585,58 @@ void controller::start_block( block_timestamp_type when,
 }
 
-block_state_ptr controller::finalize_block( const signer_callback_type& signer_callback ) {
+block_state_ptr controller::finalize_block( const signer_callback_type signer_callback ) {
```
`signer_callback` should be non-const so you can move from it.

huangminghuang committed:
fix api_tests problem

jgiszczak committed:
Remove unused producer plugin method.

huangminghuang committed:
allow yield

huangminghuang committed:
fix controller apply_block

huangminghuang committed:
fix producer_schedule_test

huangminghuang committed:
fix apply_block

huangminghuang committed:
add plumbing for rewind forkdb

jgiszczak committed:
Add remove_unsigned_head to fork database.

jgiszczak committed:
Merge branch 'jrg-EPE-1476-threaded-block-signing' of github.com:/EOSIO/eos into jrg-EPE-1476-threaded-block-signing

huangminghuang committed:
fix producer_plugin sigining failure handling

huangminghuang committed:
Merge branch 'jrg-EPE-1476-threaded-block-signing' of https://github.com/EOSIO/eos into jrg-EPE-1476-threaded-block-signing

huangminghuang committed:
hook abort_block

huangminghuang committed:
handle signing exception

jgiszczak committed:
Recover transactions from aborted unsigned head block.

huangminghuang committed:
fix abort_block

huangminghuang committed:
Merge remote-tracking branch 'origin/develop-boxed' into jrg-EPE-1476-threaded-block-signing

huangminghuang committed:
fix performance issue

huangminghuang committed:
fix abort_block

huangminghuang committed:
remove abort_block from start_block()

huangminghuang committed:
fix abort_block

jgiszczak committed:
Fix amqp plugin production paused behavior

amqp_trx_plugin calls on_block_stop when a block is aborted but when the
abort is due to a pause in production, is_producing_block is already
false by the time on_block_pause checks it.  The plugin's transaction
queue processing is inverted from the expected behavior from then on.

huangminghuang committed:
fix producer_plugin

huangminghuang committed:
Merge branch 'jrg-EPE-1476-threaded-block-signing' of https://github.com/EOSIO/eos into jrg-EPE-1476-threaded-block-signing

huangminghuang committed:
fix state history plugin

# Conflicts:
#	libraries/state_history/transaction_trace_cache.cpp

jgiszczak committed:
Add DELAY type signature provider for arbitrary signing delay in ms

Delays near or in excess of the block time can work but result in
pathological behavior.

jgiszczak committed:
Expand block tracking in amqp_trx_plugin

huangminghuang committed:
Merge branch 'develop-boxed' into jrg-EPE-1476-threaded-block-signing

huangminghuang committed:
add tests with signing delay

huangminghuang committed:
add trace_plugin_test_with_signing_delay

huangminghuang committed:
fix trace_plugin_test.py

jgiszczak committed:
Add block tracking to trace_api_plugin.

huangminghuang committed:
fix shutdown

jgiszczak committed:
Fix chain extraction unit test.

Cosmetic revisions to chain extraction, per reviewer.

huangminghuang committed:
refactor signing

huangminghuang committed:
Increase the signing delay to 200

huangminghuang committed:
add signing delay to rodeos_idle_test

huangminghuang committed:
fix fifo_trx_processing_queue

huangminghuang committed:
remove delay key signature provider

huangminghuang committed:
refactor finalize_block

huangminghuang committed:
Merge branch 'jrg-EPE-1476-threaded-block-signing' of https://github.com/EOSIO/eos into jrg-EPE-1476-threaded-block-signing

huangminghuang committed:
code clean up

huangminghuang committed:
more code clean up

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -1,3 +1,4 @@
+#include "eosio/vm/function_traits.hpp"
```
I don't see why this is needed.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2079,16 +2092,47 @@ struct controller_impl {
       if( pending ) {
          uint32_t block_num = pending->get_block_num();
-         applied_trxs = pending->extract_trx_metas();
+         ilog("abortig pending block ${block_num}", ("block_num", block_num));
```
Remove or change to `dlog`. Also aborting is misspelled.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2079,16 +2092,47 @@ struct controller_impl {
+      if (fork_db.is_signing_failed_block_head(signing_failed_blocknum)) {
+         signing_failed_blocknum = 0;
+         auto poped = pop_block();
+         ilog("abortig unsigned block ${block_num}", ("block_num", poped->block_num));
```
Remove or change to `dlog`. Also aborting is misspelled.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2079,16 +2092,47 @@ struct controller_impl {
+   void abort_unsigned_block() {
+      if( pending ) {
+         uint32_t block_num = pending->get_block_num();
+         ilog("abortig pending block ${block_num}", ("block_num", block_num));
```
Remove or change to `dlog`. Also aborting is misspelled.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2079,16 +2092,47 @@ struct controller_impl {
+      auto db_head = fork_db.head();
+      if (db_head->block && db_head->block->producer_signature == signature_type()) {
+         auto poped = pop_block();
+         ilog("abortig unsigned block ${block_num}", ("block_num", poped->block_num));
```
Remove or change to `dlog`. Also aborting is misspelled.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2588,28 +2632,44 @@ void controller::start_block( block_timestamp_type when,
+   my->pending->_block_stage = completed_block{bsp};
+
+   return async_thread_pool(my->block_sign_pool.get_executor(),
+                            [bsp, my = my.get(), block_num = bsp->block_num, digest = bsp->sig_digest(),
```
Remove `digest = bsp->sig_digest()` and instead call that from inside the lambda/thread-pool

heifner replied:
Decided safest to leave this out of the other thread so that the processing in the thread is completely stateless.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2079,16 +2092,47 @@ struct controller_impl {
+         emit( self.block_abort, block_num );
+      }
+      auto db_head = fork_db.head();
+      if (db_head->block && db_head->block->producer_signature == signature_type()) {
```
Seems like we should be able to always have this as an OR check with the `abort_block` case and then not need `abort_unsigned_block`. Either way I would like to see `abort_unsigned_block` removed.

huangminghuang committed:
address PR comments

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2588,28 +2632,44 @@ void controller::start_block( block_timestamp_type when,
 
-   auto bsp = std::make_shared<block_state>(
+   auto pfa = ab._pending_block_header_state.prev_activated_protocol_features;
+   auto pfs = my->protocol_features.get_protocol_feature_set();
```
This is an expensive copy that is not needed as `get_protocol_feature_set` returns `const protocol_feature_set&` and it is only used as a reference in this method.

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2588,28 +2632,44 @@ void controller::start_block( block_timestamp_type when,
+                            [bsp, my = my.get(), block_num = bsp->block_num, digest = bsp->sig_digest(),
+                             wtmsig_enabled = eosio::chain::detail::is_builtin_activated(
+                                 pfa, pfs, builtin_protocol_feature_t::wtmsig_block_signatures),
+                             sign = std::move(sign)]() -> std::function<void()> {
```
This lambda needs to be `mutable` for the `std::move` to work.

huangminghuang replied:
This ambda only moves `sigs` which is a local variable inside lambda; therefore, it doesn't require `mutable` to work.

heifner reviewed (changes requested):
Partial review.

heifner commented on libraries/chain/fork_database.cpp:
```diff
@@ -464,6 +464,23 @@ namespace eosio { namespace chain {
       }
    }
 
+   bool fork_database::is_signing_failed_block_head(uint32_t signing_failed_blocknum)  {
```
Seems like this should just be called `is_head_block(uint23_t blocknum)`

heifner commented on libraries/chain/fork_database.cpp:
```diff
@@ -464,6 +464,23 @@ namespace eosio { namespace chain {
+      return my->head->block && my->head->block->block_num() == signing_failed_blocknum;
+   }
+
+   void fork_database::remove_head() {
```
I think this should take the `id` or `block_num` and verify head is as expected.
Also I think this should assert if prev does not exist as it doesn't remove the head in that case.

heifner commented on libraries/state_history/include/eosio/state_history/log.hpp:
```diff
@@ -216,12 +216,14 @@ class state_history_traces_log : public state_history_log {
 
-   void block_start(uint32_t block_num) { cache.clear(); }
+   void block_start(uint32_t block_num) {
+      trace_caches.erase(block_num-2);
```
Why `-2` instead of `-1`?

heifner replied:
Is this even needed now that it is cleared in `store`?

huangminghuang replied:
`store` only clear the successful block,  not the aborted block. The trace_caches should caches at most 2 blocks, i.e. `block_num`, and `block_num-1`. Any block older than `block_num-1` can be considered obsolete.

heifner commented on plugins/amqp_trx_plugin/include/eosio/amqp_trx_plugin/fifo_trx_processing_queue.hpp:
```diff
@@ -96,7 +93,7 @@ class blocking_queue {
    bool empty() const {
       std::scoped_lock<std::mutex> lk(mtx_);
-      return queue_.empty() && !paused_;
+      return queue_.empty() && paused_ <= 0;
```
I would like to see these `paused_` changes reverted. Seems sloppy to not fix the calls so that the invariant holds instead of relaxing the invariant that unpause should only be called when paused.

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -219,6 +219,17 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
+         signatures_none,
+         signatures_pending,
+         signatures_ready
+      };
```
Can we make this an enum class

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -290,7 +301,9 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
-         _unapplied_transactions.add_aborted( chain.abort_block() );
+         _unapplied_transactions.add_aborted(chain.abort_block());
+
+         // TODO: need work on the signing failure case
```
Remove TODO. Looks to me like subjective billing does not require any changes.

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -1706,19 +1719,34 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
 
       try {
-         if( !remove_expired_trxs( preprocess_deadline ) )
+         if (!remove_expired_trxs(preprocess_deadline))
```
Please try to avoid space changes in non-modified code, it makes the diff harder to read.

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -2411,38 +2462,41 @@ void producer_plugin_impl::produce_block() {
+      auto                        on_exit      = fc::make_scoped_exit([self] {
+         self->signatures_status = signatures_ready;
+         app().post(priority::high, [self]() { self->assign_block_signatures_if_ready(); });
+      });
```
The `on_exit` doesn't add anything here. The transform can't throw except for out-of-memory. Just add these two lines of code before the `return`.

huangminghuang replied:
The `sign` can actually throw exception at least for TPM. See https://github.com/EOSIO/eos/blob/44624098a436b3694861029e71fa5b4f9f3f0742/libraries/tpm-helpers/tpm-helpers.cpp#L315

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -2411,38 +2462,41 @@ void producer_plugin_impl::produce_block() {
-        ("count",new_bs->block->transactions.size())("lib",chain.last_irreversible_block_num())("confs", new_bs->header.confirmed));
+      ("p",new_bs->header.producer)("id",new_bs->id.str().substr(8,16))
+      ("n",new_bs->block_num)("t",new_bs->header.timestamp)
+      ("count",new_bs->block->transactions.size())("lib",chain.last_irreversible_block_num())("confs", new_bs->header.confirmed));
```
I don't see any change here. Why was indent changed?

heifner commented on plugins/signature_provider_plugin/signature_provider_plugin.cpp:
```diff
@@ -90,16 +98,17 @@ signature_provider_plugin::signature_provider_plugin():my(new signature_provider
+                     "Limits the maximum time (in milliseconds) that is allowed for sending requests to a keosd "
+                     "provider for signing")("signing-delay",
+                                             boost::program_options::value<uint32_t>()->default_value(0),
+                                             "milliseconds to delay the signature signing when using default signature provider");
```
Fix formating.

huangminghuang committed:
add remove_head() back

huangminghuang committed:
more PR comment changes

huangminghuang committed:
Merge branch 'develop-boxed' into jrg-EPE-1476-threaded-block-signing

# Conflicts:
#	libraries/chain/controller.cpp

jgiszczak committed:
Merge branch 'develop-boxed' into jrg-EPE-1476-threaded-block-signing

jgiszczak committed:
Remove merge error.

jgiszczak committed:
Restore chain extraction transaction id tracking lost to merge.

jgiszczak committed:
Do not remove nonexistent signing failed block or head block

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -1602,6 +1602,17 @@ struct controller_impl {
       pending->push();
    }
 
+   void assign_signatures(block_state_ptr bsp, std::vector<signature_type>&& sigs, bool wtmsig_enabled) {
```
I don't like this does more than assign signatures. I think it needs to be renamed or the `log_irreversible()` and `emit` should be pulled out of it. 
I also don't like that the `emit(self.accepted_block, bsp);` is called from two locations now. Seems we need a concept like finalize block (although we already have that). Maybe complete_block.

jgiszczak committed:
Remove incorrect accepted_block signal emission.

Experimental, as all network-related tests fail locally due to bad firewall settings.

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -1608,17 +1619,31 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
          if( !remove_expired_trxs( preprocess_deadline ) )
             return start_block_result::exhausted;
+         
+         if (!assign_block_signatures_if_ready())
```
I think `complete_block` would be a better name.

huangminghuang committed:
code refactor based on PR review

huangminghuang committed:
remove unnecessary check in commit_block()

huangminghuang committed:
change commit_block signature

huangminghuang committed:
split commit_block functionality

huangminghuang committed:
add reserve

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -1565,47 +1565,40 @@ struct controller_impl {
-         pending.reset();
-      });
+   void create_completed_block(block_state_ptr bsp) {
+      auto reset_pending_on_exit = fc::make_scoped_exit([this] { pending.reset(); });
```
You can just add `pending.reset()` as the last line of this method. shared_ptr copy/move does not throw. `pending->push()` also does not throw.
`bsp` can be moved into `completed_block`

heifner commented on libraries/chain/controller.cpp:
```diff
@@ -2580,10 +2570,12 @@ controller::finalize_block(block_state_ptr& bsp, signer_callback_type&& sign) {
                          });
-   my->commit_block(std::move(bsp));
+   my->add_to_fork_db(bsp);
+   my->create_completed_block(bsp);
```
`bsp` can be moved here.

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -1608,17 +1619,31 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
          // limit execution of pending incoming to once per block
          size_t pending_incoming_process_limit = _unapplied_transactions.incoming_size();
 
          if( !process_unapplied_trxs( preprocess_deadline ) )
```
Need to check `complete_produced_block_if_ready` within the loop of `process_unapplied_trxs`.

huangminghuang committed:
add PR comments

huangminghuang committed:
fix uninitialized variable

huangminghuang committed:
Merge remote-tracking branch 'origin/develop-boxed' into jrg-EPE-1476-threaded-block-signing

huangminghuang committed:
return correct status from process_unapplied_trxs

huangminghuang committed:
revert maybe_producer_block change

heifner commented on plugins/trace_api_plugin/include/eosio/trace_api/chain_extraction.hpp:
```diff
@@ -71,45 +75,47 @@ class chain_extraction_impl_type {
          auto bt = create_block_trace( block_state );
+         const auto& itr = tracked_blocks.find( block_state->block_num );
 
+         if( itr == tracked_blocks.end() )
```
This should never happen. Minor thing, but an error log might be appropriate.

heifner commented on libraries/state_history/include/eosio/state_history/log.hpp:
```diff
@@ -218,12 +218,14 @@ class state_history_traces_log : public state_history_log {
 
-   void block_start(uint32_t block_num) { cache.clear(); }
+   void block_start(uint32_t block_num) {
+      trace_caches[block_num].clear();
```
This leaves the entry in the map which is a slow leak.

heifner replied:
It is cleared in store_entry

heifner commented on plugins/trace_api_plugin/include/eosio/trace_api/chain_extraction.hpp:
```diff
@@ -71,45 +75,47 @@ class chain_extraction_impl_type {
 
    void on_block_start( uint32_t block_num ) {
-      clear_caches();
+      tracked_blocks[block_num] = block_tracking{};
```
Need to listen for abort block so we can erase this otherwise this is a slow leak on every aborted block.

heifner replied:
Actually, nevermind. You can't skip a block number, obviously, so this will just wipe out the previous setting on starting the block after an abort block.

huangminghuang committed:
add error log
